core-bundle/src/Filesystem/DbafsFilesystem.php

```diff
@@ -0,0 +1,416 @@
+    {
+        $sourcePath = $this->resolvePath($source);
+        $destinationPath = $this->resolvePath($destination);
+
+        $this->adapter->move($sourcePath, $destinationPath, $this->mergeConfig($config));
+        $this->sync($sourcePath, $destinationPath);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param string|int $source
+     * @param string|int $destination
+     */
+    public function copy($source, $destination, array $config = []): void
+    {
+        $destinationPath = $this->resolvePath($destination);
+
+        $this->adapter->copy($this->resolvePath($source), $destinationPath, $this->mergeConfig($config));
+        $this->sync($destinationPath);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param string|int $path
+     */
+    public function lastModified($path): int
+    {
+        return $this->adapter->lastModified($this->resolvePath($path))->lastModified();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param string|int $path
+     */
+    public function fileSize($path): int
+    {
+        return $this->adapter->fileSize($this->resolvePath($path))->fileSize();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param string|int $path
+     */
+    public function mimeType($path): string
+    {
+        return $this->adapter->mimeType($this->resolvePath($path))->mimeType();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param string|int $path
+     */
+    public function setVisibility($path, string $visibility): void
+    {
+        $this->adapter->setVisibility($this->resolvePath($path), $visibility);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param string|int $path
+     */
+    public function visibility($path): string
+    {
+        return $this->adapter->visibility($this->resolvePath($path))->visibility();
+    }
+
+    public function setExtraMetadata(): void
+    {
+        // todo
+    }
+
+    /**
+     * @param string|int $location
+     * @phpstan-return ExtraMetadata
+     */
+    public function extraMetadata($location): array
+    {
+        $record = $this->dbafs->getRecord($this->resolvePath($location));
+
+        if (null === $record) {
+            // todo: find the right exception
+            throw new \RuntimeException('');
+        }
+
+        return $record['extra'];
+    }
+
+    /**
+     * Synchronizes the database assisted file system. If a $scope is
+     * provided only a certain file/subdirectory will be synchronized.
+     *
+     * @param string ...$scope relative paths inside the filesystem root
+     */
+    public function sync(string ...$scope): ChangeSet
+    {
+        return $this->dbafs->sync($this->adapter, ...$scope);
+    }
+
+    /**
+     * Resolve DBAFS IDs and UUIDs to paths. If a path was given, normalize it.
+     *
+     * @param string|int $identifier
+     */
+    private function resolvePath($identifier): string
+    {
+        if (is_numeric($identifier)) {
```

Comment:
This could be problematic I think. A folder with the name `123` would be inaccessible, same if someone names a file or folder afer a UUID.

How about using a value object that represents a uuid or an ID? `DbafsPath::fromUuid(…)`, `DbafsPath::fromId(…)`?

Reply:
The problem with stringable objects is that you cannot cannot pass them though abstraction that annotate `string` instead of `stringable`. Think of any regular `FilesystemOperator` or our own `VirtualFilesystem` with prefix paths.

We could maybe require a certain format that is more unlikely to match a folder name: `foo/bar/#123` or `[123]` for ID 123. :thinking: I'm not sure if UUIDs are a problem. Why would you name a folder like this? (matching an actual entry in this exact DBAFS table)

Reply:
How about a protocol prefix? `dbafs:123` and `dbafs:d6c8c82a-e588-11eb-9478-e7674e53b388`? Or `dbafs://123` and `dbafs://d6c8c82a-e588-11eb-9478-e7674e53b388`?

Reply:
Works for me, too. I'm not sure the protocol in the middle of a string looks "correct" to me, but this could only need some getting used to it.

```php
$fooStorage->extraMetadata("dbafs://$uuid");
$virtualFilesystem->read("foo/bar/dbafs://$uuid");
```

Reply:
If it can appear inside of other pahts I would prefer a folder prefix I think like `~/123` and `~/d6c8c82a-e588-11eb-9478-e7674e53b388`. This way I can also list all IDs and files by listing the contents in the `~` folder.

Reply:
Or maybe two sepcific prefixes: `~id/123` and `~uuid/d6c8c82a-e588-11eb-9478-e7674e53b388`

Reply:
As briefly discussed with @ausi on slack we could/should probably make UUIDs first class citizens of an interface that both the `DbafsFilesystem` and the `VirtualFilesystem` implements. They should be unique across all dbafs instances anyway and it's a necessity that we can reference them from content elements and other places. This would also mean you do not need a prefix for them in the `VirtualFilesystem`.

So we're just left with IDs. :thinking:

Reply:
I reworked the concept in e21aa8064f352740870ee50415f3d3d37121d29a. Now there are only UUIDs and they are "first class citizens".